command/agent/consul/check_watcher.go

```diff
@@ -0,0 +1,278 @@
+
+	// Main watch loop
+	for {
+		// Don't start watching until we actually have checks that
+		// trigger restarts.
+		for len(checks) == 0 {
+			select {
+			case c := <-w.watchCh:
+				if c.remove {
+					// should not happen
+					w.logger.Printf("[DEBUG] consul.health: told to stop watching an unwatched check: %q", c.checkID)
+				} else {
+					checks[c.checkID] = c
+					w.logger.Printf("[DEBUG] consul.health: watching alloc %q task %q check %q", c.allocID, c.taskName, c.checkName)
+
+					// Begin polling
+					if !checkTimer.Stop() {
+						<-checkTimer.C
+					}
+					checkTimer.Reset(w.pollFreq)
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+
+		// As long as there are checks to be watched, keep watching
+		for len(checks) > 0 {
+			select {
+			case c := <-w.watchCh:
+				if c.remove {
+					delete(checks, c.checkID)
+				} else {
+					checks[c.checkID] = c
+					w.logger.Printf("[DEBUG] consul.health: watching alloc %q task %q check %q", c.allocID, c.taskName, c.checkName)
+				}
+			case <-ctx.Done():
+				return
+			case <-checkTimer.C:
+				checkTimer.Reset(w.pollFreq)
+
+				// Set "now" as the point in time the following check results represent
+				now := time.Now()
+
+				results, err := w.consul.Checks()
+				if err != nil {
+					if !w.lastErr {
+						w.lastErr = true
+						w.logger.Printf("[ERR] consul.health: error retrieving health checks: %q", err)
+					}
+					continue
+				}
+
+				w.lastErr = false
+
+				// Loop over watched checks and update their status from results
+				for cid, check := range checks {
+					result, ok := results[cid]
+					if !ok {
+						// Only warn if outside grace period to avoid races with check registration
+						if now.After(check.graceUntil) {
+							w.logger.Printf("[WARN] consul.health: watched check %q (%s) not found in Consul", check.checkName, cid)
+						}
+						continue
+					}
+
+					check.update(now, result.Status)
+				}
+			}
+		}
+	}
+}
+
+// Watch a task and restart it if unhealthy.
+func (w *checkWatcher) Watch(allocID, taskName, checkID string, check *structs.ServiceCheck, restarter TaskRestarter) {
+	if !check.Watched() {
+		// Not watched, noop
+		return
+	}
+
+	c := checkRestart{
+		allocID:        allocID,
+		taskName:       taskName,
+		checkID:        checkID,
+		checkName:      check.Name,
+		task:           restarter,
+		restartDelay:   restarter.RestartDelay(),
+		interval:       check.Interval,
+		grace:          check.CheckRestart.Grace,
+		graceUntil:     time.Now().Add(check.CheckRestart.Grace),
+		timeLimit:      check.Interval * time.Duration(check.CheckRestart.Limit-1),
+		ignoreWarnings: check.CheckRestart.IgnoreWarnings,
+		logger:         w.logger,
+	}
+
+	select {
+	case w.watchCh <- &c:
+		// sent watch
+	case <-w.done:
+		// exited; nothing to do
+	}
+}
+
+// Unwatch a task.
+func (w *checkWatcher) Unwatch(cid string) {
+	c := checkRestart{
+		checkID: cid,
+		remove:  true,
+	}
+	select {
+	case w.watchCh <- &c:
```

Comment:
create an unwatchCh and just pass the cid?

Reply:
That would create a race between adding and removing watches since select cases are randomly selected.

Instead I'll make `watchCh -> watchUpdateCh` and create a small `checkRestartUpdate` struct to handle the add vs remove case.